pkg/terraformer/terraformer.go

```diff
@@ -0,0 +1,207 @@
+
+// execute is the main function of terraformer and puts all parts together (interacting with the terraform config and
+// state resources on the kubernetes cluster, executing and watching terraform calls, delegating process signals and
+// watching the state file).
+func (t *Terraformer) execute(command Command) (rErr error) {
+	sigCh := make(chan os.Signal, 1)
+	SignalNotify(sigCh, syscall.SIGINT, syscall.SIGTERM)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		select {
+		case <-sigCh:
+			t.log.Info("interrupt received")
+			cancel()
+		case <-ctx.Done():
+		}
+	}()
+
+	if err := t.EnsureTFDirs(); err != nil {
+		return fmt.Errorf("failed to create needed directories: %w", err)
+	}
+
+	if err := t.FetchConfigAndState(ctx); err != nil {
+		return err
+	}
+
+	shutdownWorker := t.StartStateUpdateWorker()
+	defer shutdownWorker()
+
+	// always try to update state ConfigMap one last time before exiting
+	defer func() {
+		err := t.TriggerAndWaitForFinalStateUpdate()
+		if rErr == nil {
+			// make sure, we don't exit with exit code 0, if we were unable to store the state
+			rErr = err
+		}
+	}()
+
+	// start file watcher for state file, that will continuously update state configmap
+	// as soon as state file changes on disk
+	shutdownFileWatcher, err := t.StartFileWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to start state file watcher: %w", err)
+	}
+
+	// stop file watcher and wait for it to be finished
+	defer shutdownFileWatcher()
+
+	// initialize terraform plugins
+	if err := t.executeTerraform(ctx, Init); err != nil {
+		return fmt.Errorf("error executing terraform %s: %w", Init, err)
+	}
+
+	// execute main terraform command
+	if err := t.executeTerraform(ctx, command); err != nil {
+		return fmt.Errorf("error executing terraform %s: %w", command, err)
+	}
+
+	if command == Validate {
+		if err := t.executeTerraform(ctx, Plan); err != nil {
+			return fmt.Errorf("error executing terraform %s: %w", Plan, err)
+		}
+	}
+
+	return nil
+}
+
+func (t *Terraformer) executeTerraform(ctx context.Context, command Command) error {
+	log := t.stepLogger("executeTerraform")
+
+	args := []string{string(command)}
+
+	switch command {
+	case Init:
+		args = append(args, "-plugin-dir="+t.paths.ProvidersDir)
+	case Plan:
+		args = append(args, "-var-file="+t.paths.VarsPath, "-parallelism=4", "-detailed-exitcode", "-state="+t.paths.StatePath)
+	case Apply:
+		args = append(args, "-var-file="+t.paths.VarsPath, "-parallelism=4", "-auto-approve", "-state="+t.paths.StatePath)
+	case Destroy:
+		args = append(args, "-var-file="+t.paths.VarsPath, "-parallelism=4", "-auto-approve", "-state="+t.paths.StatePath)
+	}
+
+	args = append(args, t.paths.ConfigDir)
+
+	log.Info("executing terraform", "command", command, "args", strings.Join(args[1:], " "))
+	tfCmd := exec.Command(TerraformBinary, args...)
+	// redirect all terraform output to stderr (same as logs)
+	tfCmd.Stdout = Stderr
+	tfCmd.Stderr = Stderr
```

Comment:
Yeah, I was also thinking about redirecting the terraformer output to the log or prefixing it somehow.
I did it this way, because we also had it like this before. But we can definitely change this later on, if we think it makes sense.
Should we keep it like this for the time being.

Reply:
Now I also remember the full reason for not doing it like that:
The terraformer pod logs are fetched by the provider extensions and they put parts of the output into the `lastError` fields (which in turn is used to calculate error codes). 
So I wanted to be a bit more cautious here, as this might 1) look really bad in the dashboard (if the logs JSON fields are not stripped) and 2) might even break our error code detection in an unforeseeable way.
So, we can definitely revise that idea, but I guess we will also need some changes in `g/g/ext/terraformer` with it, too.